waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,381 @@
+            (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                            {:current-time current-time
+                             :expiry-time not-on-or-after
+                             :status 400})))
+        email (first (get attrs "email"))
+        ; https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/technical-reference/the-role-of-claims
+        upn (first (get attrs "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn"))
+        name-id-value (get name-id :value)
+        saml-principal (or upn name-id-value)
+        {:keys [authorization/principal authorization/user]} (auth/auth-params-map saml-principal)
+        saml-principal' (if (and email (= principal user)) email saml-principal)
+        saml-auth-data (utils/map->base-64-string
+                         {:not-on-or-after not-on-or-after :redirect-url request-url :saml-principal saml-principal'}
+                         password)]
+    {:body (render-authenticated-redirect-template {:auth-redirect-uri (str scheme "://" host auth-redirect-endpoint)
+                                                    :saml-auth-data saml-auth-data})
+     :status 200}))
+
+(defn saml-auth-redirect-handler
+  "Endpoint for POST back to Waiter with SAML authentication data. If data is still valid,
+   add waiter authentication cookie and redirect back to the original user app."
+  [{:keys [password]} request]
+  {:pre [(not-empty password)]}
+  (if-let [saml-auth-data (get-in (ring-params/params-request request) [:form-params "saml-auth-data"])]
+    (let [{:keys [not-on-or-after redirect-url saml-principal]}
+          (try
+            (utils/base-64-string->map saml-auth-data password)
+            (catch Exception e
+              (throw (ex-info "Could not parse saml-auth-data." {:inner-exception e
+                                                                 :saml-auth-data saml-auth-data
+                                                                 :status 400} e))))
+          current-time (t/now)
+          _ (when-not (t/before? current-time not-on-or-after)
+              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                              {:current-time current-time
+                               :expiry-time not-on-or-after
+                               :status 400})))
+          auth-cookie-expiry-date (t/min-date not-on-or-after (t/plus current-time (t/days 1)))
+          auth-cookie-age-in-seconds (-> current-time
+                                       (t/interval auth-cookie-expiry-date)
+                                       t/in-seconds)
+          {:keys [authorization/principal authorization/user] :as auth-params-map}
+          (auth/auth-params-map saml-principal)]
+      (auth/handle-request-auth (constantly {:body ""
+                                             :headers {"location" redirect-url}
+                                             :status 303})
```

Comment:
It would be nice to know the count and rate of saml auth redirects

Reply:
are you referring to reading the cookie and allowing auth if successful, I think we do not need metrics for that, but feel free to add if you like.